spec/java/util/StringBuilder.main.lsl

```diff
@@ -0,0 +1,1022 @@
+            newStr[i] = currentChar_2;
+            currentIndex += 1;
+        }
+        else
+        {
+            val index: int = i - dstOffset;
+            val currentChar_3: char = action CALL_METHOD(this.storage, "charAt", [index]);
+            newStr[i] = currentChar_3;
+        }
+    }
+
+
+    proc _substring(start: int, end: int): String
+    {
+        _checkRangeSIOOBE(start, end, this.length);
+        val sizeNewString: int = end - start;
+        val newStr: array<char> = action ARRAY_NEW("char", sizeNewString);
+
+        var i: int = 0;
+        action LOOP_FOR(i, start, end, +1, _newSubString_loop(i, newStr));
+        result = action OBJECT_TO_STRING(newStr);
+    }
+
+
+    @Phantom proc _newSubString_loop (i: int, newStr: array<char>): void
+    {
+        newStr[i] = action CALL_METHOD(this.storage, "charAt", [i]);
+    }
+
+    // constructors
+
+    constructor *.StringBuilder (@target self: StringBuilder)
+    {
+
+    }
+
+
+    constructor *.StringBuilder (@target self: StringBuilder, seq: CharSequence)
+    {
+        if (seq == null)
+            _throwNPE();
+
+        _appendCharSequence(seq);
+    }
+
+
+    constructor *.StringBuilder (@target self: StringBuilder, str: String)
+    {
+        if (str == null)
+            _throwNPE();
+
+        _appendString(str);
+    }
+
+
+    constructor *.StringBuilder (@target self: StringBuilder, capacity: int)
+    {
+
+    }
+
+
+    // methods
+
+    fun *.append (@target self: StringBuilder, seq: CharSequence): StringBuilder
+    {
+        _appendCharSequence(seq);
+
+        result = self;
+    }
+
+
+    fun *.append (@target self: StringBuilder, seq: CharSequence, start: int, end: int): StringBuilder
+    {
+        if (seq == null)
+            seq = "null";
+        val seqLength: int = action CALL_METHOD(seq, "length", []);
+        _checkRange(start, end, seqLength);
+        this.length += end - start;
+        var i: int = 0;
+        action LOOP_FOR(i, start, end, +1, _appendCharSequence_loop(i, seq));
+        result = self;
+    }
+
+
+    fun *.append (@target self: StringBuilder, obj: Object): StringBuilder
+    {
+        if (obj == null)
+        {
+            this.storage += "null";
+            this.length += 4;
+        }
+        else
+        {
+            this.storage += action OBJECT_TO_STRING(obj);
+            this.length = action CALL_METHOD(this.storage, "length", []);
+        }
+        result = self;
+    }
+
+
+    fun *.append (@target self: StringBuilder, str: String): StringBuilder
+    {
+        _appendString(str);
+
+        result = self;
+    }
+
+
+    fun *.append (@target self: StringBuilder, sb: StringBuffer): StringBuilder
+    {
+        if (sb == null)
+        {
+            this.storage += "null";
+            this.length += 4;
+        }
+        else
+        {
+            // That's right for StringBuffer ?
+            this.storage += action OBJECT_TO_STRING(sb);
```

Comment:
@dog-m In such way ?

```
            if (sb has StringBufferAutomaton)
            {
                this.storage += StringBufferAutomaton(sb).storage;
                this.length += StringBufferAutomaton(sb).length;
            }
```

But it doesn't work - we don't have StringBufferAutomaton at this moment; ofcourse I can describe it, but it can lead to merge conflict with Valentin;

Reply:
I suggest adding it but commented-out. I hope it will be "enabled" back into existence when `StringBuffer` be ready for merge.